heudiconv/bids.py

```diff
@@ -487,3 +510,505 @@ def convert_sid_bids(subject_id):
+
+    Parameters:
+    ----------
+    json_file : str
+
+    Returns:
+    -------
+    str with "ShimSetting" value
+    """
+    data = load_json(json_file)
+    try:
+        shims = data[SHIM_KEY]
+    except KeyError as e:
+        lgr.error('File %s does not have "ShimSetting".'
+                  'Please use a different "matching_parameters" in your heuristic file', json_file)
+        raise KeyError
+    return shims
+
+
+def find_fmap_groups(fmap_dir):
+    """
+    Finds the different fmap groups in a fmap directory.
+    By groups here we mean fmaps that are intended to go together
+    (with reversed PE polarity, magnitude/phase, etc.)
+
+    Parameters:
+    ----------
+    fmap_dir : str or os.path
+        path to the session folder (or to the subject folder, if there are no
+        sessions).
+
+    Returns:
+    -------
+    fmap_groups : dict
+        key: prefix common to the group (e.g. no "dir" entity, "_phase"/"_magnitude", ...)
+        value: list of all fmap paths in the group
+    """
+    if op.basename(fmap_dir) != 'fmap':
+        lgr.error('%s is not a fieldmap folder', fmap_dir)
+
+    # Get a list of all fmap json files in the session:
+    fmap_jsons = sorted(glob(op.join(fmap_dir, '*.json')))
+
+    # RegEx to remove fmap-specific substrings from fmap file names
+    # "_phase[1,2]", "_magnitude[1,2]", "_phasediff", "_dir-<label>", ...
+    fmap_regex = re.compile(
+        '(_dir-[0-9,a-z,A-Z]*)*'  # for pepolar case
+        '(_phase[12])*'  # for phase images
+        '(_phasediff)*'  # for phasediff images
+        '(_magnitude[12])*'  # for magnitude images
+        '(_fieldmap)*'  # for actual fieldmap images
+    )
+
+    # Find the unique prefixes ('splitext' removes the extension):
+    prefixes = sorted(
+        set(fmap_regex.sub('', remove_suffix(op.basename(fm), '.json')) for fm in fmap_jsons)
+    )
+    fmap_groups = OrderedDict()
+    for k in prefixes:
+        fmap_groups[k] = [
+            fm for fm in fmap_jsons if fmap_regex.sub('', remove_suffix(op.basename(fm), '.json')) == k
+        ]
+    return fmap_groups
+
+
+def get_key_info_for_fmap_assignment(json_file, matching_parameter):
+    """
+    Gets key information needed to assign fmaps to other modalities.
+
+    Parameters:
+    ----------
+    json_file : str or os.path
+        path to the json file
+    matching_parameter : str in AllowedFmapParameterMatching
+        matching_parameter that will be used to match runs
+
+    Returns:
+    -------
+    key_info : dict
+        part of the json file that will need to match between the fmap and
+        the other image
+    """
+    if not op.exists(json_file):
+        raise FileNotFoundError(
+            errno.ENOENT, os.strerror(errno.ENOENT), json_file
+        )
+    if matching_parameter not in AllowedFmapParameterMatching:
+        raise ValueError(
+            "Fmap matching_parameter %s not allowed." % matching_parameter
+        )
```

Comment:
there is 4 identical spots verifying value to be the one expected.  Should probably be done once only at the point of where that value is coming in (heuristic?)

Reply:
Sure, I can do it only once. However, I wanted to make each function self-sufficient, so that it could be re-used if needed.  (Disclaimer: I have no Computer Science/Programming training, so I really don't know what the convention is for these things...:blush: )

Reply:
we all learn as we go/do more coding ;)  typically if there is repeated / copy-pasted code, it is a sign for either needing to take that code into a function for reuse, or as in this case AFAIK, move checking somewhere closer to where the value could "enter" (be specified by the user) the code path.

Reply:
Addressed in d53e50b7b9ddbbeded6765c353fd12bf882664fb